etna/models/nn/patchts.py

```diff
@@ -0,0 +1,379 @@
+                                        nn.Linear(in_features=self.hidden_size * self.max_patch_num, out_features=1))
+
+        self.lr = lr
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+
+    def forward(self, x: PatchTSBatch, *args, **kwargs):  # type: ignore
+        """Forward pass.
+
+        Parameters
+        ----------
+        x:
+            batch of data
+
+        Returns
+        -------
+        :
+            forecast with shape (batch_size, decoder_length, 1)
+        """
+        encoder_real = x["encoder_real"].float()  # (batch_size, encoder_length-1, input_size)
+        decoder_real = x["decoder_real"].float()  # (batch_size, decoder_length, input_size)
+        decoder_length = decoder_real.shape[1]
+        outputs = []
+        x = encoder_real
+        for i in range(decoder_length):
+            pred = self._get_prediction(x)
+            outputs.append(pred)
+            x = torch.cat((x[:, 1:, :], torch.unsqueeze(pred, dim=1)), dim=1)
+
+        forecast = torch.cat(outputs, dim=1)
+        forecast = torch.unsqueeze(forecast, dim=2)
+
+        return forecast
+
+    def _get_prediction(self, x: torch.Tensor) -> torch.Tensor:
+        x = x.permute(0, 2, 1)  # (batch_size, input_size, encoder_length + decoder_length - 1)
+        # do patching
+        x = x.unfold(dimension=-1, size=self.patch_len,
+                     step=self.stride)  # (batch_size, input_size, patch_num, patch_len)
+
+        y = self.model(x)
+        y = y.permute(1, 0, 2)  # (batch_size, hidden_size, patch_num)
+
+        return self.projection(y)  # (batch_size, 1)
+
+    def step(self, batch: PatchTSBatch, *args, **kwargs):  # type: ignore
+        """Step for loss computation for training or validation.
+
+        Parameters
+        ----------
+        batch:
+            batch of data
+
+        Returns
+        -------
+        :
+            loss, true_target, prediction_target
+        """
+        encoder_real = batch["encoder_real"].float()  # (batch_size, encoder_length-1, input_size)
+        decoder_real = batch["decoder_real"].float()  # (batch_size, decoder_length, input_size)
+
+        decoder_target = batch["decoder_target"].float()  # (batch_size, decoder_length, 1)
+
+        decoder_length = decoder_real.shape[1]
+
+        outputs = []
+        x = encoder_real
+        for i in range(decoder_length):
+            pred = self._get_prediction(x)
+            outputs.append(pred)
+            x = torch.cat((x[:, 1:, :], torch.unsqueeze(decoder_real[:, i, :], dim=1)), dim=1)
+
+        target_prediction = torch.cat(outputs, dim=1)
+        target_prediction = torch.unsqueeze(target_prediction, dim=2)
+
+        loss = self.loss(target_prediction, decoder_target)
+        return loss, decoder_target, target_prediction
+
+    def make_samples(self, df: pd.DataFrame, encoder_length: int, decoder_length: int) -> Iterator[dict]:
```

Comment:
we should add test cases for `make_samples` for logic checks and for regression testing